Python/compile.c

```diff
@@ -6175,20 +6175,44 @@ compiler_pattern_mapping(struct compiler *c, pattern_ty p, pattern_context *pc)
     }
     // Collect all of the keys into a tuple for MATCH_KEYS and
     // COPY_DICT_WITHOUT_KEYS. They can either be dotted names or literals:
+
+    // will hold only Constant_kind keys
+    PyObject *seen = PySet_New(NULL);
+    if (seen == NULL) {
+        return -1;
+    }
+
     for (Py_ssize_t i = 0; i < size; i++) {
         expr_ty key = asdl_seq_GET(keys, i);
         if (key == NULL) {
             const char *e = "can't use NULL keys in MatchMapping "
                             "(set 'rest' parameter instead)";
             SET_LOC(c, ((pattern_ty) asdl_seq_GET(patterns, i)));
+            Py_DECREF(seen);
             return compiler_error(c, e);
         }
+
+        if (key->kind == Constant_kind) {
+            if (PySet_Contains(seen, key->v.Constant.value)) {
```

Comment:
This needs to be checked for errors:

```suggestion
            int in_seen = PySet_Contains(seen, key->v.Constant.value);
            if (in_seen < 0) {
                Py_DECREF(seen);
                return 0;
            }
            if (in_seen) {
```